custom-local-authenticator/src/main/java/org/wso2/custom/authenticator/local/CustomLocalAuthenticator.java

```diff
@@ -0,0 +1,162 @@
+import org.wso2.carbon.identity.application.authentication.framework.exception.InvalidCredentialsException;
+import org.wso2.carbon.identity.application.authentication.framework.model.AuthenticatedUser;
+import org.wso2.carbon.identity.application.authentication.framework.util.FrameworkUtils;
+import org.wso2.carbon.user.core.UserStoreManager;
+import org.wso2.carbon.user.core.common.AbstractUserStoreManager;
+import org.wso2.carbon.utils.multitenancy.MultitenantUtils;
+
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+
+/**
+ * Custom Local Authenticator that allow a user to authenticate only if the has the REQUIRED_ROLE.
+ */
+public class CustomLocalAuthenticator extends AbstractApplicationAuthenticator
+        implements LocalApplicationAuthenticator {
+
+    private static final long serialVersionUID = 4345354156955223654L;
+    private static final Log log = LogFactory.getLog(CustomLocalAuthenticator.class);
+
+    private static final String AUTHENTICATOR_NAME = "CustomLocalAuthenticator";
+    private static final String AUTHENTICATOR_FRIENDLY_NAME = "Custom Local Authenticator";
+    private static final String USER_NAME = "username";
+    private static final String PASSWORD = "password";
+    private static final String REQUIRED_ROLE = "customRole";
+    private static final String OIDC = "oidc";
+
+    @Override
+    protected void initiateAuthenticationRequest(HttpServletRequest request,
+                                                 HttpServletResponse response,
+                                                 AuthenticationContext context)
+            throws AuthenticationFailedException {
+
+        log.info("This is the initiateAuthenticationRequest method");
+        String loginPage = ConfigurationFacade.getInstance().getAuthenticationEndpointURL();
+        // This is the default WSO2 IS login page. If you can create your custom login page you can use
+        // that instead.
+        String queryParams =
+                FrameworkUtils.getQueryStringWithFrameworkContextId(context.getQueryParams(),
+                        context.getCallerSessionKey(),
+                        context.getContextIdentifier());
+        try {
+            String retryParam = "";
+
+            if (context.isRetrying()) {
+                retryParam = "&authFailure=true&authFailureMsg=login.fail.message";
+            }
+
+            response.sendRedirect(response.encodeRedirectURL(loginPage + ("?" + queryParams)) +
+                    "&authenticators=BasicAuthenticator:" + "LOCAL" + retryParam);
+        } catch (IOException e) {
+            throw new AuthenticationFailedException(e.getMessage(), e);
+        }
+    }
+
+    /**
+     * This method is used to process the authentication response.
+     */
+    @Override
+    protected void processAuthenticationResponse(HttpServletRequest request,
```

Comment:
This authentication only checks the OIDC request state and whether user has the role? there will be no username password validaiton?

Reply:
User can use any preferred authentication mechanism. I've added a comment.